src/Microsoft.Extensions.ML/ServiceCollectionExtensions.cs

```diff
@@ -19,27 +20,50 @@ public static class ServiceCollectionExtensions
         /// <param name="services">
         /// The <see cref="IServiceCollection "/> to add services to.
         /// </param>
+        /// <param name="implementationFactory">
+        /// The factory that creates the <see cref="PredictionEnginePoolBuilder{TData, TPrediction}"/>.
+        /// </param>
         /// <returns>
         /// The <see cref="PredictionEnginePoolBuilder{TData, TPrediction}"/> that was added to the collection.
         /// </returns>
         public static PredictionEnginePoolBuilder<TData, TPrediction> AddPredictionEnginePool<TData, TPrediction>(
-            this IServiceCollection services)
+            this IServiceCollection services,
+            Func<IServiceProvider, PredictionEnginePool<TData, TPrediction>> implementationFactory)
             where TData : class
             where TPrediction : class, new()
         {
-            services.AddPredictionEngineServices<TData, TPrediction>();
+            services
+                .AddPrerequisiteServices()
+                .AddSingleton<PredictionEnginePool<TData, TPrediction>, PredictionEnginePool<TData, TPrediction>>(implementationFactory);
             return new PredictionEnginePoolBuilder<TData, TPrediction>(services);
         }
 
-        internal static IServiceCollection AddPredictionEngineServices<TData, TPrediction>(
+        /// <summary>
+        /// Adds a <see cref="PredictionEnginePoolBuilder{TData, TPrediction}"/> to the service collection.
+        /// </summary>
+        /// <param name="services">
+        /// The <see cref="IServiceCollection "/> to add services to.
+        /// </param>
+        /// <returns>
+        /// The <see cref="PredictionEnginePoolBuilder{TData, TPrediction}"/> that was added to the collection.
+        /// </returns>
+        public static PredictionEnginePoolBuilder<TData, TPrediction> AddPredictionEnginePool<TData, TPrediction>(
             this IServiceCollection services)
             where TData : class
             where TPrediction : class, new()
         {
-            services.AddLogging();
-            services.AddOptions();
-            services.TryAddEnumerable(ServiceDescriptor.Singleton<IPostConfigureOptions<MLOptions>, PostMLContextOptionsConfiguration>());
-            services.AddSingleton<PredictionEnginePool<TData, TPrediction>, PredictionEnginePool<TData, TPrediction>>();
+            services
+                .AddPrerequisiteServices()
+                .AddSingleton<PredictionEnginePool<TData, TPrediction>, PredictionEnginePool<TData, TPrediction>>();
+            return new PredictionEnginePoolBuilder<TData, TPrediction>(services);
+        }
+
+        public static IServiceCollection AddPrerequisiteServices(this IServiceCollection services)
```

Comment:
`AddPrerequisiteServices` is too vague, since this is an extension on `IServiceCollection`, people won't know what kind of pre-reqs are being added.

What do you think of renaming it to `AddPredictionEnginePoolPrerequisites`?

Reply:
I felt a little uncomfortable with that name as it could imply to the user that they must call this method _before_ calling `AddPredictionEnginePool`.  I wanted to think of a name that indicates that it _only_ does something _less_ than `AddPredictionEnginePool`.  So I came up with `AddPredictionEnginePoolConfigOnly` given that the services being added appear to be configuration required for `PredictionEnginePool` (i.e., logging, options, post-configure options).

Grammatically, it makes sense also.  In `AddPredictionEnginePool`, "PredictionEnginePool" is the noun. In `AddPredictionEnginePoolConfigOnly`, "PredictionEnginePool" becomes an adjective while "Config" becomes the noun. "Only" is an adverb in this case.  "Add" in both cases is of course the verb.

Let me know what you think.  I want to be careful with the naming on this as it will be there forever once set. 🙂